src/torchpme/tuning/tuner.py

```diff
@@ -0,0 +1,325 @@
+            positions,
+            cutoff,
+            neighbor_indices,
+            neighbor_distances,
+            True,
+        )
+
+    def tune(self, accuracy: float = 1e-3) -> tuple[list[float], list[float]]:
+        """
+        Estimate the error and timing for each parameter set. Only parameters for
+        which the error is smaller than the accuracy are timed, the others' timing is
+        set to infinity.
+
+        :param accuracy: a float, the desired accuracy
+        :return: a list of errors and a list of timings
+        """
+        if not isinstance(accuracy, float):
+            raise ValueError(f"'{accuracy}' is not a float.")
+        smearing = self.estimate_smearing(accuracy)
+        param_errors = []
+        param_timings = []
+        for param in self.params:
+            error = self.error_bounds(smearing=smearing, cutoff=self.cutoff, **param)  #  type: ignore[call-arg]
+            param_errors.append(float(error))
+            if error > accuracy:
+                param_timings.append(float("inf"))
+                continue
+
+            param_timings.append(self._timing(smearing, param))
+
+        return param_errors, param_timings
+
+    def _timing(self, smearing: float, k_space_params: dict):
+        calculator = self.calculator(
+            potential=CoulombPotential(
+                # exponent=self.exponent,  # but only exponent = 1 is supported
+                smearing=smearing,
+            ),
+            **k_space_params,
+        )
+
+        return self.time_func(calculator)
+
+
+class TuningTimings(torch.nn.Module):
+    """
+    Class for timing a calculator.
+
+    The class estimates the average execution time of a given calculater after several
+    warmup runs. The class takes the information of the structure that one wants to
+    benchmark on, and the configuration of the timing process as inputs.
+
+    :param charges: atomic charges
+    :param cell: single tensor of shape (3, 3), describing the bounding
+    :param positions: single tensor of shape (``len(charges), 3``) containing the
+        Cartesian positions of all point charges in the system.
+    :param cutoff: real space cutoff, serves as a hyperparameter here.
+    :param neighbor_indices: torch.Tensor with the ``i,j`` indices of neighbors for
+        which the potential should be computed in real space.
+    :param neighbor_distances: torch.Tensor with the pair distances of the neighbors for
+        which the potential should be computed in real space.
+    :param n_repeat: number of times to repeat to estimate the average timing
+    :param n_warmup: number of warmup runs
+    :param run_backward: whether to run the backward pass
+    """
+
+    def __init__(
+        self,
+        charges: torch.Tensor,
+        cell: torch.Tensor,
+        positions: torch.Tensor,
+        cutoff: float,
+        neighbor_indices: Optional[torch.Tensor] = None,
+        neighbor_distances: Optional[torch.Tensor] = None,
```

Comment:
Set them as required arguments. We do the same for the actual calculations and let user decide how they want to compute their neighbors.